streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyOptimizerImpl.java

```diff
@@ -0,0 +1,379 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.streams.errors.TopologyException;
+import org.apache.kafka.streams.processor.FailOnInvalidTimestamp;
+import org.apache.kafka.streams.processor.StreamPartitioner;
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+
+/**
+ * The {@code TopologyOptimizer} used to optimize topology built using the DSL.
+ * This implementation traverses the graph and makes calls to the {@link InternalTopologyBuilder} instead
+ * of calls from inheritors of {@link AbstractStream}
+ */
+@SuppressWarnings("unchecked")
+public class TopologyOptimizerImpl implements TopologyOptimizer {
+
+    private final NodeIdComparator nodeIdComparator = new NodeIdComparator();
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopologyOptimizerImpl.class);
+
+    @Override
+    public void optimize(final StreamsTopologyGraph topologyGraph,
+                         final InternalTopologyBuilder internalTopologyBuilder) {
+
+        final PriorityQueue<StreamsGraphNode> graphNodePriorityQueue = new PriorityQueue<>(5, nodeIdComparator);
+
+        graphNodePriorityQueue.offer(topologyGraph.getRoot());
+
+        LOG.debug("Root node {} descendants {}", topologyGraph.getRoot(), topologyGraph.getRoot().descendants);
+
+        while (!graphNodePriorityQueue.isEmpty()) {
+            final StreamsGraphNode streamGraphNode = graphNodePriorityQueue.remove();
+
+            if (LOG.isDebugEnabled()) {
+                List<StreamsGraphNode> allDescendantNodes = new ArrayList<>();
+                if (streamGraphNode.topologyNodeType == TopologyNodeType.SOURCE) {
+                    allDescendantNodes = getAllDescendants(streamGraphNode, new ArrayList<StreamsGraphNode>());
+                }
+                LOG.debug("{} descendant nodes {}", streamGraphNode, allDescendantNodes);
+            }
+
+            buildAndMaybeOptimize(internalTopologyBuilder, streamGraphNode);
+            for (StreamsGraphNode graphNode : streamGraphNode.getDescendants()) {
+                graphNodePriorityQueue.offer(graphNode);
+            }
+        }
+    }
+
+    private void buildAndMaybeOptimize(final InternalTopologyBuilder internalTopologyBuilder,
```

Comment:
Correct me if I'm wrong: my understanding is that the optimization execute by traversing the graph and build based on each single GN. In that case how is that different with our current mechanism? For example, if we want to add an optimization rule to merge multiple repartition topics, how could that rule be executed based on this model??

Reply:
TLDR; for now we'll have hard-coded rules that will be used to build up different data structures in the `StreamsToplogyGraphImpl`, then the rules are going to be applied once the graph is traversed in the optimizer phase.  An example of what I'm referring to follows here.

While the optimizer will work on one graph node at a time, the key point is decisions for optimizations won't occur isolation.  Under our current model, we add an operation to the topology without any context.  We make one - N number calls to the InternalTopologyBuilder adding all the required parts for the topology, but we lose the contextual information at that point.  While we register the name of the previous operation of a parent operation, this only serves to wire to topology together; there is no context as to what the last action is, or for that matter what operations the DSL added N steps prior.

In this model as we add a node to the graph, the `StreamsTopologyGraphImpl` connects the current node with its predecessor in the `setPredecessor` call, and likewise, the current node is connected with its predecessor as a descendant node (`StreamsTopologyGraphImpl` lines 71 -72).  This connectivity allows for traversing in either direction to inspect the meta-data of a given node enabling for making any optimization decisions.


Secondly, when adding a node to the graph, we can use the type of the node as a hint that there is an opportunity to optimize.  While we won't perform the optimization at this point, we'll record the fact that an operation from the DSL is eligible for optimization. For example, consider the following code:

```
KStream<String, String> inputStream = builder.stream("input-topic");

KStream<String, String> mappedStream = inputStream.map(new KeyValueMapper<String, String, KeyValue<? extends String, ? extends String>>() {
            @Override
            public KeyValue<? extends String, ? extends String> apply(String key, String value) {

                return KeyValue.pair(key.substring(0, 3), value);
            }
        });

mappedStream.groupByKey().windowedBy(TimeWindows.of(5000)).count().toStream().to("count-one-out");
mappedStream.groupByKey().windowedBy(TimeWindows.of(10000)).count().toStream().to("count-two-out");
mappedStream.groupByKey().windowedBy(TimeWindows.of(15000)).count().toStream().to("count-three-out");
```

The node representing the `inputStream.map` gets stored as a key in `Map<StreamsGraphNode, Set<StreamsGraphNode>> repartitioningNodeToRepartitioned` and the child nodes of `mappedStream`, the `groupByKey` nodes, end up stored in the `Set<StreamGraphNode>`.

Now going back to where this fits into the `ToplogyOptimizer` as we traverse the graph to build the final topology, we can check each node to see if there have been any optimizations recorded for this particular node.  It's at that point that we'll apply any of the optimizations.  In this case, we'd check the `repartitioningNodeToRepartitioned` mentioned above, and we would do a manual `through` operation to repartition the `mappedStream` and then update the child nodes of the `groupByKey` to use the new information when added to the topology via the InternalStreamsBuilder.

During this first pass, only building the graph and applying a couple of hard-coded rules are going to be done.  In another pass, we'll develop a mechanism for plugging in different rules.